core/src/test/scala/org/apache/spark/storage/DiskBlockObjectWriterCloseMethodBenchmark.scala

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.storage
+
+import org.apache.spark.{SparkContext, SparkEnv}
+import org.apache.spark.benchmark.{Benchmark, BenchmarkBase}
+import org.apache.spark.executor.ShuffleWriteMetrics
+import org.apache.spark.internal.config
+import org.apache.spark.serializer.SerializerInstance
+import org.apache.spark.util.collection.{PartitionedPairBuffer, WritablePartitionedIterator}
+
+/**
+ * Benchmark for `close` vs `revertPartialWritesAndClose` method in `DiskBlockObjectWriter`.
+ * To run this benchmark:
+ * {{{
+ *   1. without sbt:
+ *      bin/spark-submit --class <this class> <spark core test jar>
+ *   2. build/sbt "core/test:runMain <this class>"
+ *   3. generate result:
+ *      SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "core/test:runMain <this class>"
+ *      Results will be written to
+ *      "benchmarks/DiskBlockObjectWriterCloseMethodBenchmark-results.txt".
+ * }}}
+ */
+object DiskBlockObjectWriterCloseMethodBenchmark extends BenchmarkBase {
+
+  val sc = new SparkContext(master = "local", appName = "test")
+
+  private[this] def spillWithClose(
+      iterator: WritablePartitionedIterator[Int, Int],
+      blockManager: BlockManager,
+      serInstance: SerializerInstance,
+      fileBufferSize: Int,
+      serializerBatchSize: Long,
+      revertAndClose: Boolean): Unit = {
+    val (blockId, file) = blockManager.diskBlockManager.createTempShuffleBlock()
+
+    var objectsWritten: Long = 0
+    val spillMetrics: ShuffleWriteMetrics = new ShuffleWriteMetrics
+    val writer: DiskBlockObjectWriter =
+      blockManager.getDiskWriter(blockId, file, serInstance, fileBufferSize, spillMetrics)
+
+    def flush(): Unit = {
+      writer.commitAndGet()
+      objectsWritten = 0
+    }
+
+    while (iterator.hasNext) {
+      iterator.writeNext(writer)
+      objectsWritten += 1
+
+      if (objectsWritten == serializerBatchSize) {
+        flush()
+      }
+    }
+    if (objectsWritten > 0) {
+      flush()
+      writer.close()
+    } else if (revertAndClose) {
+      writer.revertPartialWritesAndClose()
+    } else {
+      writer.close()
+    }
+  }
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
```

Comment:
It is difficult to compare benchmark directly between `writer.close()` and `writer.revertPartialWritesAndClose()`, so I extracted the main code part of `ExternalSorter.spill` to compare the performance of `writer.close()` and `writer.revertPartialWritesAndClose()`

@mridulm @Ngone51